lib/Bio/EnsEMBL/DataCheck/Checks/CheckSpeciesTrees.pm

```diff
@@ -0,0 +1,90 @@
+
+Licensed under the Apache License, Version 2.0 (the 'License');
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an 'AS IS' BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+=cut
+
+package Bio::EnsEMBL::DataCheck::Checks::CheckSpeciesTrees;
+
+use warnings;
+use strict;
+
+use Moose;
+use Test::More;
+use Bio::EnsEMBL::DataCheck::Test::DataCheck;
+use Bio::EnsEMBL::Utils::SqlHelper;
+use Data::Dumper;
+
+extends 'Bio::EnsEMBL::DataCheck::DbCheck';
+
+use constant {
+  NAME           => 'CheckSpeciesTrees',
+  DESCRIPTION    => 'The expected number of species trees have been merged',
+  GROUPS         => ['compara', 'compara_gene_trees', 'compara_genome_alignments'],
+  DATACHECK_TYPE => 'critical',
+  DB_TYPES       => ['compara'],
+  TABLES         => ['method_link_species_set', 'species_set', 'species_tree_root']
+};
+
+sub tests {
+  my ($self) = @_;
+  my $dba    = $self->dba;
+  my $helper = $dba->dbc->sql_helper;
+  my @method_links = qw(CACTUS_HAL EPO EPO_EXTENDED PECAN PROTEIN_TREES NC_TREES SPECIES_TREE);
+
+  my $expected_tree_count;
+
+  foreach my $method_link_type ( @method_links ) {
+
+    my $mlsss = $self->dba->get_MethodLinkSpeciesSetAdaptor->fetch_all_by_method_link_type($method_link_type);
+    # Only check from the method_links that have mlsss there are other datachecks to check if mlsss are correct
+    next if scalar(@$mlsss) == 0;
+
+    foreach my $mlss ( @$mlsss ) {
+
+      my $mlss_id   = $mlss->dbID;
+      my $mlss_name = $mlss->name;
+
+      my $sql = qq/
+        SELECT COUNT(*)
+          FROM species_tree_root
+        WHERE method_link_species_set_id = $mlss_id
+      /;
+
+      $expected_tree_count += $helper->execute_single_result(-SQL => $sql);
+
+      my $desc_1 = "The species tree for $mlss_id ($mlss_name) is present as expected";
+      is_rows_nonzero($dba, $sql, $desc_1);
+    }
+  }
+
+  # Check that all the species_trees correspond to a method_link_species_set that should have a species tree
+  my $desc_2 = "All the species_trees with corresponding method_link_species_sets are expected";
+  my $row_count_sql = "SELECT COUNT(*) FROM species_tree_root";
+  cmp_rows($dba, $row_count_sql, "==", $expected_tree_count, $desc_2);
+
+  my $species_tree_root_constraint = q/
+    method_link_id IN (
+      SELECT method_link_id FROM method_link
+      WHERE
+        method_link_id < 100 AND
+        (class LIKE 'GenomicAlignTree%' OR class LIKE '%multiple_alignment')
+    )
+  /;
+  fk($dba, 'species_tree_root', 'method_link_species_set_id', 'method_link_species_set', 'method_link_species_set_id', $species_tree_root_constraint);
```

Comment:
I know this is an exact copy of the original `fk` check, but looking into it, I wonder if that's right.

My understanding of how `fk` works is that the additional constraint should apply to the first table, not the second one. See https://github.com/Ensembl/ensembl-datacheck/blob/release/100/lib/Bio/EnsEMBL/DataCheck/Test/DataCheck.pm#L336-L378 : `fk` does `WHERE t1.$col1 IS NOT NULL AND t2.$col2 IS NULL`, so all the rows in `t2` have to be NULLs, so none can't match the condition. In summary, this `fk` (and other similar ones in ForeignKeysCompara) do nothing ? Can anyone double-check this ?

Reply:
I think you are right, that it is not right. I am trying to get my head around the fact that the `LEFT JOIN` is on `method_link_species_set_id` in both tables, the `IS NULL` / `IS NOT NULL` would not work because the `method_link_id` is from the second table, which has to be `IS NULL`- which means it is not there? Is that similar to your thinking? The check in this case looks like:
`    SELECT * FROM
    species_tree_root t1 LEFT JOIN method_link_species_set t2 ON t1.method_link_species_set_id = t2.method_link_species_set_id;
    WHERE t1.method_link_species_set_id IS NULL AND t2.method_link_species_set_id IS NOT NULL AND 
    method_link_id IN (
      SELECT method_link_id FROM method_link
      WHERE
        method_link_id < 100 AND
        (class LIKE 'GenomicAlignTree%' OR class LIKE '%multiple_alignment')
    );`

Reply:
Yep that's my thinking.

Ah right, good idea. It's actually easier with the actual query:
```
ensro@mysql-ens-compara-prod-1.ebi.ac.uk:4485/ensembl_compara_101 > EXPLAIN SELECT * FROM species_tree_root t1 LEFT JOIN method_link_species_set t2 ON t1.method_link_species_set_id = t2.method_link_species_set_id WHERE t1.method_link_species_set_id IS NULL AND t2.method_link_species_set_id IS NOT NULL AND method_link_id IN ( SELECT method_link_id FROM method_link WHERE method_link_id < 100 AND (class LIKE 'GenomicAlignTree%' OR class LIKE '%multiple_alignment') );
+----+-------------+-------+------+---------------+------+---------+------+------+------------------+
| id | select_type | table | type | possible_keys | key  | key_len | ref  | rows | Extra            |
+----+-------------+-------+------+---------------+------+---------+------+------+------------------+
|  1 | SIMPLE      | NULL  | NULL | NULL          | NULL | NULL    | NULL | NULL | Impossible WHERE |
+----+-------------+-------+------+---------------+------+---------+------+------+------------------+
1 row in set (0.033 sec)
```

MySQL is pretty clear about what it thinks of the `WHERE` :)

Reply:
If we come back to the original HCs:
- https://github.com/Ensembl/ensj-healthcheck/blob/release/100/src/org/ensembl/healthcheck/testcase/compara/ForeignKeyMLSSIdGenomic.java#L45-L59 and https://github.com/Ensembl/ensj-healthcheck/blob/release/100/src/org/ensembl/healthcheck/testcase/compara/ForeignKeyMLSSIdHomology.java#L59-L60
- https://github.com/Ensembl/ensj-healthcheck/blob/release/100/src/org/ensembl/healthcheck/testcase/compara/AbstractMLSSIdToData.java#L30-L41

Two sorts of checks were performed on the `species_tree_root` table:
1. All the MLSSs that match the constraint have some data in the `species_tree_root` table
2. `species_tree_root.method_link_species_set_id` are valid mlss_ids

Check n°2 is actually a real SQL foreign key check that is encoded in our schema and captured by ForeignKeysCompara: https://github.com/Ensembl/ensembl-compara/blob/release/100/sql/table.sql#L469

Check n°1 is actually what you have implemented, but with method_link types instead of classes, which is fine.

So I think you could delete this block and it'd be OK ?
```suggestion
```

And we need to create a ticket to fix all the `fk` calls of ForeignKeysCompara that have a condition on the wrong table.